controllers/workflow_controller.go

```diff
@@ -0,0 +1,454 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+
+	wfv1 "github.com/argoproj/argo-workflows/v3/pkg/apis/workflow/v1alpha1"
+	"github.com/go-logr/logr"
+	addonapiv1 "github.com/keikoproj/addon-manager/api/addon"
+	addonv1 "github.com/keikoproj/addon-manager/api/addon/v1alpha1"
+	pkgaddon "github.com/keikoproj/addon-manager/pkg/addon"
+	"github.com/keikoproj/addon-manager/pkg/common"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	wfcontroller = "workflow-controller"
+)
+
+type wfreconcile struct {
+	client       client.Client
+	log          logr.Logger
+	versionCache pkgaddon.VersionCacheClient
+}
+
+func NewWFController(mgr manager.Manager, stopChan <-chan struct{}, addonversioncache pkgaddon.VersionCacheClient) (controller.Controller, error) {
+	r := &wfreconcile{
+		client:       mgr.GetClient(),
+		log:          ctrl.Log.WithName(wfcontroller),
+		versionCache: addonversioncache,
+	}
+
+	c, err := controller.New(wfcontroller, mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return nil, err
+	}
+
+	err = c.Watch(&source.Kind{Type: &wfv1.Workflow{}}, r.enqueueRequestForOwner())
+	if err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+func (r *wfreconcile) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
```

Comment:
I appreciate this new direction, using the reconcile mechanism is correct way IMO. 👍